File: pe-math/src/main/java/org/saar/maths/transform/Scale.java
package org.saar.maths.transform;

import org.joml.Vector3f;
import org.joml.Vector3fc;
import org.jproperty.value.ObservableValueBase;
import org.saar.maths.utils.Vector3;

public class Scale extends ObservableValueBase<Vector3fc> implements ReadonlyScale {

    private final Vector3f value = Vector3.create();

    private Scale(Vector3fc value) {
        this.value.set(value);
    }

    public static Scale of(float x, float y, float z) {
        return new Scale(Vector3.of(x, y, z));
    }

    public static Scale create() {
        return new Scale(Vector3.of(1));
    }

    private Vector3fc copyValue() {
        return Vector3.of(this.value);
    }

    public void scale(float scale) {
        final Vector3fc old = copyValue();
        this.value.mul(scale);
        onChange(old);
    }

    public void scale(Vector3fc scale) {
        final Vector3fc old = copyValue();
        this.value.mul(scale);
        onChange(old);
    }

    public void scale(float x, float y, float z) {
        final Vector3fc old = copyValue();
        this.value.mul(x, y, z);
        onChange(old);
    }

    public void set(ReadonlyScale scale) {
        final Vector3fc old = copyValue();
        this.value.set(scale.getValue());
        onChange(old);
    }

    public void set(Vector3fc value) {
        final Vector3fc old = copyValue();
        this.value.set(value);
        onChange(old);
    }

    public void set(float x, float y, float z) {
        final Vector3fc old = copyValue();
        this.value.set(x, y, z);
        onChange(old);
    }

    public void set(float value) {
        final Vector3fc old = copyValue();
        this.value.set(value);
        onChange(old);
    }

    private void onChange(Vector3fc old) {
        if (!old.equals(getValue())) {
            fireChangeEvent(old);
        }
    }

    @Override
    public Vector3fc getValue() {
        return this.value;
    }

    @Override
    public String toString() {
        return "Scale{" + getValue() + '}';
    }
}

File: pe-math/src/main/kotlin/org/saar/maths/transform/Scale.kt
package org.saar.maths.transform

import org.joml.Vector3fc
import org.jproperty.value.ObservableValueBase
import org.saar.maths.utils.Vector3

class Scale private constructor(value: Vector3fc) : ObservableValueBase<Vector3fc>(), ReadonlyScale {

    private val value = Vector3.of(value)

    private fun copyValue(): Vector3fc {
        return Vector3.of(this.value)
    }

    fun scale(scale: Float) {
        val old = copyValue()
        this.value.mul(scale)
        onChange(old)
    }

    fun scale(scale: Vector3fc) {
        val old = copyValue()
        this.value.mul(scale)
        onChange(old)
    }

    fun scale(x: Float, y: Float, z: Float) {
        val old = copyValue()
        this.value.mul(x, y, z)
        onChange(old)
    }

    fun set(scale: ReadonlyScale) {
        val old = copyValue()
        this.value.set(scale.value)
        onChange(old)
    }

    fun set(value: Vector3fc) {
        val old = copyValue()
        this.value.set(value)
        onChange(old)
    }

    fun set(x: Float, y: Float, z: Float) {
        val old = copyValue()
        this.value.set(x, y, z)
        onChange(old)
    }

    fun set(value: Float) {
        val old = copyValue()
        this.value.set(value)
        onChange(old)
    }

    private fun onChange(old: Vector3fc) {
        if (old != this.value) {
            fireChangeEvent(old)
        }
    }

    override fun getValue(): Vector3fc {
        return this.value
    }

    override fun toString(): String {
        return "Scale{" + this.value + '}'
    }

    companion object {
        fun of(x: Float, y: Float, z: Float): Scale {
            return Scale(Vector3.of(x, y, z))
        }

        @JvmStatic
        fun create(): Scale {
            return Scale(Vector3.of(1f))
        }
    }
}
